### PR TITLE
Add AckableChatRoom sample

### PR DESCRIPTION
**Pull request by lizh-lixiang**
This is a simple chat room with acks for private messages (user to user).
@vicancy This is the complete project "AckableChatRoom" which uses ackhandler.

### CODE REVIEW

File: samples/AckableChatRoom/Hub/AckableChatRoom.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{
+    public class AckableChatRoom : Hub
+    {
+        private readonly IAckHandler _ackHandler;
+
+        public AckableChatRoom(IAckHandler ackHandler)
+        {
+            _ackHandler = ackHandler;
+        }
+        public void BroadcastMessage(string name, string message)
```
Comment (lizh-lixiang) [inline samples/AckableChatRoom/Hub/AckableChatRoom.cs]: Blankline

File: samples/AckableChatRoom/wwwroot/index.html [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                return v.toString(16);
+            });
+        }
+
+        // Html encode message.
+        function encodedMessage(message) {
+            return message.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
+        }
+
+        function createPublicMessage(encodedName, encodedMsg, encodeMsgId) {
```
Comment (lizh-lixiang) [inline samples/AckableChatRoom/wwwroot/index.html]: Can merge createPublicMessage and createUserMessage together.

File: samples/AckableChatRoom/wwwroot/index.html [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                    `<div class="message-content pull-right">${encodedMsg}<div>`;
+            } else {
+                entry.innerHTML =
+                    `<div class="message-avatar pull-left" id="${encodeMsgId}">${encodedName}</div>` +
+                    `<div class="message-content pull-left">${encodedMsg}<div>`;
+            }
+            return entry;
+        }
+
+        function createUserMessage(msgId, sourceName, msg) {
```
Comment (lizh-lixiang) [inline samples/AckableChatRoom/wwwroot/index.html]: `createUserMessage(messageId, sender, messageText)`

File: samples/AckableChatRoom/wwwroot/index.html [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            <textarea id="message" style="width: 100%; padding: 5px 10px; border-style: hidden;"
+                      placeholder="Type message and press Enter to send..."></textarea>
+        </div>
+        <div style="overflow: auto; border-style: ridge; border-top-style: hidden;">
+            <button class="btn-success pull-right" id="sendmessage">Send</button>
+        </div>
+        <div>
+            <textarea id="targetName" style="width: 100%; padding: 5px 10px; border-style: hidden;"
+                      placeholder="Type targetName"></textarea>
+            <button class="btn-success pull-right" id="sendUserMessage">sendUserMessage</button>
```
Comment (lizh-lixiang) [inline samples/AckableChatRoom/wwwroot/index.html]: Better naming

File: samples/AckableChatRoom/wwwroot/index.html [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    <meta http-equiv="Pragma" content="no-cache" />
+    <meta http-equiv="Expires" content="0" />
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@3.3.7/dist/css/bootstrap.min.css" rel="stylesheet" />
+    <link href="css/site.css" rel="stylesheet" />
+    <title>Azure SignalR Group Chat With Ack</title>
+</head>
+
+<body>
+    <h2 class="text-center" style="margin-top: 0; padding-top: 30px; padding-bottom: 30px;">
+        Azure SignalR Group Chat With Ack
```
Comment (lizh-lixiang) [inline samples/AckableChatRoom/wwwroot/index.html]: Azure SignalR Single Chat

File: samples/AckableChatRoom/wwwroot/css/site.css [hunk 6]
```diff
@@ -0,0 +1,82 @@
+﻿/*html, body {
```
Comment (lizh-lixiang) [inline samples/AckableChatRoom/wwwroot/css/site.css]: Some can be removed

File: .gitignore [hunk 7]
```diff
@@ -9,4 +9,4 @@ asrs.log.txt
 **/global.json
 PublishProfiles/
 **.log.txt
-**.log
\ No newline at end of file
+**.log
```
Comment (lizh-lixiang) [inline .gitignore]: revert this global file

File: samples/AckableChatRoom/AckHandler/AckHandler.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            var id = Guid.NewGuid().ToString();
+            var tcs = new TaskCompletionSource<AckResult>(TaskCreationOptions.RunContinuationsAsynchronously);
+            _handlers.TryAdd(id, (tcs, DateTime.UtcNow));
+            return new AckInfo(id, tcs.Task);
+        }
+
+        public void Ack(string id)
+        {
+            if (_handlers.TryGetValue(id, out var res))
```
Comment (vicancy) [inline samples/AckableChatRoom/AckHandler/AckHandler.cs]: >TryGetValue [](start = 26, length = 11)

`TryRemove`?

File: samples/AckableChatRoom/AckHandler/AckHandler.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            {
+                res.Item1.TrySetResult(AckResult.Success);
+            }
+            else
+            {
+                throw new Exception("AckId not found");
```
Comment (vicancy) [inline samples/AckableChatRoom/AckHandler/AckHandler.cs]: >Exception [](start = 26, length = 9)

no need to throw?

File: samples/AckableChatRoom/AckHandler/AckHandler.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        private void CheckAcks()
+        {
+            foreach (var pair in _handlers)
+            {
+                var elapsed = DateTime.UtcNow - pair.Value.Item2;
+                if (elapsed > _ackThreshold)
+                {
+                    pair.Value.Item1.TrySetResult(AckResult.TimeOut);
```
Comment (vicancy) [inline samples/AckableChatRoom/AckHandler/AckHandler.cs]: >TrySetResult [](start = 37, length = 12)

is SetException better?

File: samples/AckableChatRoom/Hub/AckableChatRoom.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            _ackHandler = ackHandler;
+        }
+
+        public void BroadcastMessage(string name, string message)
+        {
+            Clients.All.SendAsync("broadcastMessage", name, message);
+        }
+
+        //  Complete the task specified by the ackId.
+        public void AckResponse(string ackId)
```
Comment (vicancy) [inline samples/AckableChatRoom/Hub/AckableChatRoom.cs]: >AckResponse [](start = 20, length = 11)

naming: is `Ack` enough?

File: samples/AckableChatRoom/Hub/AckableChatRoom.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            _ackHandler.Ack(ackId);
+        }
+
+        //  Send the message to the receiver
+        public async Task<string> SendUserMessage(string id, string receiver, string message)
+        {
+            //  Create a task and wait for the receiver client to complete it.
+            var ackInfo = _ackHandler.CreateAck();
+            var sender = Context.User.FindFirst(ClaimTypes.NameIdentifier).Value;
```
Comment (vicancy) [inline samples/AckableChatRoom/Hub/AckableChatRoom.cs]: >sender [](start = 16, length = 6)

move to the .ctor and having sender as a local field

File: samples/AckableChatRoom/Hub/AckableChatRoom.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            await Clients.User(receiver).SendAsync("displayUserMessage", id, sender, message, ackInfo.AckId);
```
Comment (vicancy) [inline samples/AckableChatRoom/Hub/AckableChatRoom.cs]: >id [](start = 73, length = 2)

what is this `id` for? Is it not necessary?

File: samples/AckableChatRoom/Hub/AckableChatRoom.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+            //  Return the task result to the client.
+            return (await ackInfo.AckTask).ToString();
+        }
+
+        // Send a customized receipt to the message sender.
+        public async Task<string> SendUserAck(string msgId, string sourceName, string message)
```
Comment (vicancy) [inline samples/AckableChatRoom/Hub/AckableChatRoom.cs]: >SendUserAck [](start = 34, length = 11)

I still feel that this one can be merged with `SendUserMessage` with a `MessageType` field added for the receiver client

File: samples/AckableChatRoom/wwwroot/index.html [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+        </div>
+    </div>
+
+    <!--Reference the SignalR library. -->
+    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/@aspnet/signalr@1.1.0/dist/browser/signalr.min.js">
+    </script>
+
+    <!--Add script to update the page and send messages.-->
+    <script type="text/javascript">
+        var username = generateRandomName();
```
Comment (vicancy) [inline samples/AckableChatRoom/wwwroot/index.html]: >generateRandomName [](start = 23, length = 18)

move all inside the `DOMContentLoaded` callback?

File: samples/AckableChatRoom/wwwroot/index.html [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+            function onConnectionError(error) {
+                if (error && error.message) {
+                    console.error(error.message);
+                }
+                const modal = document.getElementById('myModal');
+                modal.classList.add('in');
+                modal.style = 'display: block;';
+            }
+
+            bindConnectionMessage(connection);
```
Comment (vicancy) [inline samples/AckableChatRoom/wwwroot/index.html]: >bindConnectionMessage [](start = 12, length = 21)

let's follow the airbnb JavaScript coding convention [https://github.com/airbnb/javascript#functions](https://github.com/airbnb/javascript#functions)

File: samples/AckableChatRoom/wwwroot/index.html [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+                //  Add the message to the screen
+                const displayPublicMessage = function (name, message) {
+                    const messageEntry = createNewMessage(name, encodedMessage(message), generateGuid());
+
+                    addNewMessageToScreen(messageEntry);
+                };
+
+                const displayUserMessage = function (id, name, message, ackId) {
+                    const messageEntry = createNewMessage(name, encodedMessage(message), id);
+                    messageEntry.addEventListener('click', function (event) {
```
Comment (vicancy) [inline samples/AckableChatRoom/wwwroot/index.html]: >addEventListener [](start = 33, length = 16)

no need?

File: samples/AckableChatRoom/Hub/AckableChatRoom.cs (hunk 14 above)
Comment (lizh-lixiang) [reply samples/AckableChatRoom/Hub/AckableChatRoom.cs]: Merged

File: samples/AckableChatRoom/AckHandler/AckHandler.cs (hunk 10 above)
Comment (lizh-lixiang) [reply samples/AckableChatRoom/AckHandler/AckHandler.cs]: Now use SetException().
A little question:
I could see the exception in the back end.
But in the front end, only "Failed to invoke..." was displayed, no detailed exception messages.
Should I use try...catch to return the exception message?

File: samples/AckableChatRoom/wwwroot/index.html (hunk 17 above)
Comment (lizh-lixiang) [reply samples/AckableChatRoom/wwwroot/index.html]: Why?
Then how should receiver ack the message.
I think we have two choices:
Mark all as read when visibility change
Or
Mark one as read by click (Current)

File: samples/AckableChatRoom/AckHandler/AckHandler.cs [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        private void CheckAcks()
+        {
+            foreach (var pair in _handlers)
+            {
+                var elapsed = DateTime.UtcNow - pair.Value.Item2;
+                if (elapsed > _ackThreshold)
+                {
+                    pair.Value.Item1.TrySetException(new Exception("Ack time out"));
```
Comment (vicancy) [inline samples/AckableChatRoom/AckHandler/AckHandler.cs]: throw a more specific exception, e.g. `TimeOutException`

File: samples/AckableChatRoom/Startup.cs [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddMvc();
+            services.AddSignalR()
+                    .AddAzureSignalR(options =>
+                    {
+                        options.ClaimsProvider = context => new[]
+                        {
+                            new Claim(ClaimTypes.NameIdentifier, context.Request.Query["username"])
```
Comment (vicancy) [inline samples/AckableChatRoom/Startup.cs]: Provide a comment that this is a tricky part for sample purpose, for PROD usage we suggest using Authorization

File: samples/AckableChatRoom/wwwroot/index.html [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+            do {
+                username = prompt(promptMessage, username);
+                if (!username || username.startsWith('_') || username.indexOf('<') > -1 || username.indexOf(
+                    '>') > -1) {
+                    username = '';
+                    promptMessage = 'Invalid input. Enter your name:';
+                }
+            } while (!username)
+
+            const connection = new signalR.HubConnectionBuilder()
```
Comment (vicancy) [inline samples/AckableChatRoom/wwwroot/index.html]: move down to below the function definition?

File: samples/AckableChatRoom/wwwroot/index.html (hunk 17 above)
Comment (vicancy) [reply samples/AckableChatRoom/wwwroot/index.html]: click does not sound a natural operation...how about chat sample is focused?

File: samples/AckableChatRoom/Hub/AckableChatSampleHub.cs [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        //  Complete the task specified by the ackId.
+        public void Ack(string ackId)
+        {
+            _ackHandler.Ack(ackId);
+        }
+
+        //  Send the messageContent to the receiver
+        public async Task<string> SendUserMessage(string messageId, string sender, string receiver, string messageContent)
```
Comment (lizh-lixiang) [inline samples/AckableChatRoom/Hub/AckableChatSampleHub.cs]: The sender here means the message sender.
If it is a response('read') message, the sender isn't `Context.userIdentifier`.

File: samples/AckableChatRoom/wwwroot/index.html (hunk 17 above)
Comment (lizh-lixiang) [reply samples/AckableChatRoom/wwwroot/index.html]: Yes
Only when the message is in the view, a response will be sent.
Checks when user scrolls (focused first).

**Review by vicancy — APPROVED** [review]

**Pull request merged**